prov/efa/src/efa_mr.c

```diff
@@ -239,12 +239,18 @@ static int efa_mr_hmem_setup(struct efa_mr *efa_mr,
 				return err;
 			}
 		}
-		err = cuda_dev_register((struct fi_mr_attr *)attr, &efa_mr->peer.device.cuda);
-		if (err) {
-			EFA_WARN(FI_LOG_MR,
-				 "Unable to register handle for GPU memory. err: %d buf: %p len: %zu\n",
-				 err, attr->mr_iov->iov_base, attr->mr_iov->iov_len);
-			return err;
+
+		/* When cudaMemcpy is allowed (cuda_xfer_setting==CUDA_XFER_ENABLED), use cudaMemcpy (which does not need registration)
+		 * Otherwise, use gdrcopy and register the memory with gdrcopy.
+		 */
+		if (cuda_get_xfer_setting()!=CUDA_XFER_ENABLED) {
```

Comment:
This implies the user should not set `FI_HMEM_CUDA_ENABLE_XFER=1` if they want to use GDRCopy, but I was under the impression `FI_HMEM_CUDA_ENABLE_XFER=1` would still allow GDRCopy if it's available. Does this mean only `CUDA_XFER_UNSPECIFIED` would be the only way to use GDRCopy?

Reply:
Yes. cudaMemcpy should be used when it is allowed to be used. Gdrcopy is an alternative.